vcfconcat.c

```diff
@@ -683,7 +686,23 @@ static void concat(args_t *args)
                     args->seen_seq[chr_id] = 1;
                     prev_chr_id = chr_id;
 
-                    if ( vcf_write_line(args->out_fh, &fp->line)!=0 ) error("\nFailed to write %"PRIu64" bytes\n", (uint64_t)fp->line.l);
+                    if (!args->out_fh->idx) {
+                        if ( vcf_write_line(args->out_fh, &fp->line)!=0 ) error("\nFailed to write %"PRIu64" bytes\n", (uint64_t)fp->line.l);
+                    } else {
+                        // need to find reflen
+                        str = end + 1; // should be the beginning of the ref
+                        while (*str && *str != '\t') str++; // Get past the ID entry
+                        char *beg = str + 1;
+                        str++;
+
+                        while (*str && *str != '\t') {
+                            str++; // Get past the REF entry
+                        }
+
+                        if (vcf_write_line_with_index(args->out_fh, args->out_hdr, &fp->line, chr_id, tmp.s, pos, (str - beg)) < 0) {
```

Comment:
I'm not sure if this matters or not for something deliberately optimised like concat, but there is some oddity in VCF here.

The internal data structures have:

```
typedef struct bcf1_t {
    hts_pos_t pos;  // POS
    hts_pos_t rlen; // length of REF
```

On the face of it, that leads us all to assume rlen is simply reference length and can be gleaned trivially from strlen(ref) after parsing.  However I don't think that's always true and I think the header file documentation is outdated.

Specifically if there is an INFO/END tag it can override rlen, as seen here:  https://github.com/samtools/htslib/blob/develop/vcf.c#L3601-L3614

The `vcf_parse_info` function is quite heavyweight and totally goes against the grain of a fast concat loop doing nothing more than read/writes, but then again if we want to get an index which is the same that we'd get from `bcftools index` then I think that's what we have to do.  Sadly!  I suppose minimally we could do a strstr to look for `END=` and if it's found anywhere then do full parsing to check if it's END and not something else and to parse it.  It's a lot more work though.

Maybe @pd3 can advise.

I know this data exists in theory, but is it the sort of thing that only occurs in very specific examples (eg structural variation callers or gVCF), and if so would it be acceptable to simply document that `--write-index` is only valid for data from SNP and short-indel callers?